llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1336,33 +1355,86 @@ void AsmPrinter::emitFrameAlloc(const MachineInstr &MI) {
-      .encode();
+  return {MBB.isReturnBlock(), !MBB.empty() && TII->isTailCall(MBB.back()),
+          MBB.isEHPad(), const_cast<MachineBasicBlock &>(MBB).canFallThrough(),
+          !MBB.empty() && MBB.rbegin()->isIndirectBranch()};
+}
+
+static void emitPGOFunctionData(const MachineFunction &MF,
+                                MCStreamer &OutStreamer) {
+  if (PgoBBAddrMapFeatures.isSet(PGOMapFeaturesEnum::FuncEntryCnt)) {
+    OutStreamer.AddComment("function entry count");
+    OutStreamer.emitULEB128IntValue(
+        llvm::transformOptional(MF.getFunction().getEntryCount(),
+                                std::mem_fn(&Function::ProfileCount::getCount))
+            .value_or(0));
+  }
+}
+
+static void emitPGOBBData(const MachineFunction &MF,
+                          const MachineBasicBlock &MBB,
+                          const MachineBlockFrequencyInfo &MBFI,
+                          const MachineBranchProbabilityInfo &MBPI,
+                          MCStreamer &OutStreamer) {
+  bool UseBrProb = PgoBBAddrMapFeatures.isSet(PGOMapFeaturesEnum::BrProb);
+  unsigned SuccCount;
+  auto SuccsType = object::PGOBBAddrMap::BBEntry::SuccessorsType::None;
+  if (UseBrProb) {
+    SuccCount = MBB.succ_size();
+    SuccsType = object::PGOBBAddrMap::BBEntry::getSuccessorsType(SuccCount);
+  }
+
+  // Must place metadata before extra info
+  OutStreamer.emitULEB128IntValue(object::PGOBBAddrMap::BBEntry::encodeMD(
+      getBBAddrMapMetadata(MBB), SuccsType));
+
+  if (PgoBBAddrMapFeatures.isSet(PGOMapFeaturesEnum::BBFreq)) {
+    OutStreamer.AddComment("basic block frequency");
+    OutStreamer.emitULEB128IntValue(MBFI.getBlockFreq(&MBB).getFrequency());
+  }
+
+  if (UseBrProb) {
+    if (SuccsType == object::PGOBBAddrMap::BBEntry::SuccessorsType::Multiple) {
+      OutStreamer.AddComment("basic block successor count");
+      OutStreamer.emitULEB128IntValue(SuccCount);
+    }
+    for (const MachineBasicBlock *SuccMBB : MBB.successors()) {
+      OutStreamer.AddComment("successor BB ID");
+      OutStreamer.emitULEB128IntValue(SuccMBB->getBBID()->BaseID);
+      OutStreamer.AddComment("successor branch probability");
+      OutStreamer.emitULEB128IntValue(
+          MBPI.getEdgeProbability(&MBB, SuccMBB).getNumerator());
+    }
+  }
 }
 
 void AsmPrinter::emitBBAddrMapSection(const MachineFunction &MF) {
-  MCSection *BBAddrMapSection =
-      getObjFileLowering().getBBAddrMapSection(*MF.getSection());
+  unsigned PGOFeatures = PgoBBAddrMapFeatures.getBits();
+  bool UsePGOExtension = PGOFeatures != 0;
+
+  MCSection *BBAddrMapSection = getObjFileLowering().getBBAddrMapSection(
+      *MF.getSection(), UsePGOExtension);
   assert(BBAddrMapSection && ".llvm_bb_addr_map section is not initialized.");
 
   const MCSymbol *FunctionSymbol = getFunctionBegin();
+  uint8_t BBAddrMapVersion = OutStreamer->getContext().getBBAddrMapVersion();
+  assert((UsePGOExtension ? BBAddrMapVersion >= 2 : true) &&
```

Comment:
FWIW as far as I can tell `OutStreamer->getContext().getBBAddrMapVersion()` is never set anywhere in the whole LLVM codebase so is always 2 in practice. @rlavaee would it make sense to remove this field and accessor and just put a constant somewhere instead? (out of scope for this diff of course, should do that separately)

Reply:
Also noticed this and would be curious if it's overridden or could just be a constant.